grad detach_ only when it has grad_fn in zero_grad call

zhaojuanmao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#41283 grad detach_ only when it has grad_fn in zero_grad call**

in optimizer.zero_grad(), detach_ is useful to avoid memory leak only when grad has grad_fn, so add check to call grad.detach_ only when the grad has grad_fn in zero_grad() function

Differential Revision: [D22487315](https://our.internmc.facebook.com/intern/diff/D22487315/)

zhaojuanmao committed:
grad detach_ only when it has grad_fn in zero_grad call

in optimizer.zero_grad(), detach_ is useful to avoid memory leak only when grad has grad_fn, so add check to call grad.detach_ only when the grad has grad_fn in zero_grad() function

Differential Revision: [D22487315](https://our.internmc.facebook.com/intern/diff/D22487315/)

[ghstack-poisoned]

albanD commented on torch/optim/optimizer.py:
```diff
@@ -168,7 +168,8 @@ def zero_grad(self):
                 if p.grad is not None:
-                    p.grad.detach_()
+                    if p.grad.grad_fn is not None:
+                        p.grad.detach_()
```
On top of removing `.grad_fn`, it also removes the `.requires_grad` flag. I think you want to add a `p.grad.requires_grad_(False)` even when there is no grad_fn.

zhaojuanmao committed:
Update on "grad detach_ only when it has grad_fn in zero_grad call"


in optimizer.zero_grad(), detach_ is useful to avoid memory leak only when grad has grad_fn, so add check to call grad.detach_ only when the grad has grad_fn in zero_grad() function

Differential Revision: [D22487315](https://our.internmc.facebook.com/intern/diff/D22487315/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "grad detach_ only when it has grad_fn in zero_grad call"


in optimizer.zero_grad(), detach_ is useful to avoid memory leak only when grad has grad_fn, so add check to call grad.detach_ only when the grad has grad_fn in zero_grad() function

Differential Revision: [D22487315](https://our.internmc.facebook.com/intern/diff/D22487315/)

[ghstack-poisoned]

albanD reviewed (approved):
I feel like this should work fine.
Make sure that all CI is green in case we missed something.

zhaojuanmao committed:
Update on "grad detach_ only when it has grad_fn in zero_grad call"


in optimizer.zero_grad(), detach_ is useful to avoid memory leak only when grad has grad_fn, so add check to call grad.detach_ only when the grad has grad_fn in zero_grad() function

Differential Revision: [D22487315](https://our.internmc.facebook.com/intern/diff/D22487315/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "grad detach_ only when it has grad_fn in zero_grad call"


in optimizer.zero_grad(), detach_ is useful to avoid memory leak only when grad has grad_fn, so add check to call grad.detach_ only when the grad has grad_fn in zero_grad() function

Differential Revision: [D22487315](https://our.internmc.facebook.com/intern/diff/D22487315/)

[ghstack-poisoned]

zhaojuanmao commented:
CI is green, landing

zhaojuanmao committed:
Update on "grad detach_ only when it has grad_fn in zero_grad call"


in optimizer.zero_grad(), detach_ is useful to avoid memory leak only when grad has grad_fn, so add check to call grad.detach_ only when the grad has grad_fn in zero_grad() function

Differential Revision: [D22487315](https://our.internmc.facebook.com/intern/diff/D22487315/)

[ghstack-poisoned]
